test/chef_secrets_fd_tests.erl

```diff
@@ -11,16 +11,58 @@ read_returns_an_error_if_var_does_exist_test() ->
 read_returns_ejson_read_from_fd_referenced_in_env_var_test() ->
     Path = filename:join([code:priv_dir(chef_secrets), "../test/json_secrets_file.json"]),
     CredJson = os:cmd(io_lib:format("veil-dump-secrets ~s", [Path])),
-    Fd = get_fd_from_content(CredJson),
+    %% Note: OTP 21 detects leaked filehandles and GC's them fast, so we must hang onto the Fh for the duration of the test
+    {Fh, Fd} = get_fd_from_content(CredJson),
     os:putenv("CHEF_SECRETS_FD", erlang:integer_to_list(Fd)),
     {ok, Ejson} = chef_secrets_fd:read([]),
     Actual = ej:get([<<"postgresql">>, <<"db_superuser_password">>], Ejson),
     ?assertMatch(<<"37f5c43dd8b", _/binary>>, Actual),
-    os:unsetenv("CHEF_SECRETS_FD").
+    os:unsetenv("CHEF_SECRETS_FD"),
+    file:close(Fh).
 
 get_fd_from_content(Data) ->
     {A, B, C} = erlang:timestamp(),
     TmpFile = io_lib:format("/tmp/chef_secrets~w~w~w", [A, B, C]),
     ok = file:write_file(TmpFile, Data),
-    {ok, {file_descriptor, _, {_Port, Fd}}} = file:open(TmpFile, [read, raw]),
+    {ok, IoDevice} = file:open(TmpFile, [read, raw]),
+    Fd = get_fd_from_io_device(TmpFile, IoDevice),
+    {IoDevice, Fd}.
+
+
+-ifdef(open_returns_map).
+get_fd_from_io_device(Name, _) ->
+    %% I am ashamed of this hack.  The problem is that OTP-21 rewrote
+    %% raw files to use a nif. This has the effect of burying the file
+    %% descriptor deep in an opaque reference, and I've not spotted a
+    %% sane way to extract that info.
```

Comment:
Oh wow, you aren't kidding, depending on the code path you take I'm not even sure that information makes it outo fo the C nif back into erlang.